cpu/efm32/periph/i2c.c

```diff
@@ -54,28 +56,21 @@ static mutex_t i2c_lock[I2C_NUMOF];
  */
 static int _transfer(i2c_t dev, I2C_TransferSeq_TypeDef *transfer)
 {
-    bool busy = true;
-    uint32_t timeout = I2C_TIMEOUT;
+    uint32_t counter = EFM32_I2C_TIMEOUT;
 
-    /* start the i2c transaction */
-    i2c_progress[dev] = I2C_TransferInit(i2c_config[dev].dev, transfer);
+    /* start transfer (using interrupts) and wait for it to complete or
+       timeout */
+    NVIC_ClearPendingIRQ(i2c_config[dev].irq);
+    NVIC_EnableIRQ(i2c_config[dev].irq);
 
-    /* the transfer progresses via the interrupt handler */
-    while (busy) {
-        unsigned int cpsr = irq_disable();
+    i2c_progress[dev] = I2C_TransferInit(i2c_config[dev].dev, transfer);
 
-        if (i2c_progress[dev] == i2cTransferInProgress && timeout--) {
-            cortexm_sleep_until_event();
-        }
-        else {
-            busy = false;
-        }
+    while (i2c_progress[dev] == i2cTransferInProgress && counter--) {}
```

Comment:
alternatively you could also use a mutex to signal the end of the transfer and `xtimer_mutex_lock_timeout()` to wait for it - not sure how much overhead that is.

Reply:
I can try that. Maybe make it optional so that an end-user could select it or not (Kconfig?).

Reply:
A mutex with `xtimer_mutex_lock_timeout` seems to work well. Overhead in size (for STK3200, Cortex M0+) for `examples/default` with `i2c_scan`:

* `xtimer_mutex_lock_timeout` + `pm_block`: 30496 bytes
* `xtimer_mutex_lock_timeout`: 30480 bytes
* this PR: 28816 bytes
* this PR + dummy `xtimer_mutex_lock_timeout`: 30504 bytes
* old code: 28752

`pm_block` is necessary to ensure idle thread doesn't save too much energy ;-)

Reply:
See 9f472c3 for the implementation of the first result.

Reply:
Unfortunately, this approach doesn't seem to work for the SLTB001a. That one uses I2C in `board.c`, which now relies on the scheduler to be active already.

Reply:
> Unfortunately, this approach doesn't seem to work for the SLTB001a. That one uses I2C in board.c, which now relies on the scheduler to be active already.

Sounds like a bug :wink: 
Why does it need I2C in `board_init()`?

Reply:
Oops, I meant xtimer instead of scheduler ;-)

It needs this because the SLTB001a contains another MCU (power-interrupt-controller) that is responsible for enabling power lines. It needs to enable all the sensors before `auto_init` kicks in. There is currently no `board_init` that runs after xtimer has started.

Reply:
Something like 2914d45 would 'fix' it. Or if #15002 gets merged.

Reply:
Could that be initialized by `auto_init`? If it sits on top of auto_init, it would run before all the other sensors that depend on it are initialized.

That would keep the periph driver implementations simple in that they can always assume to be called from thread context (not from ISR, not before the scheduler runs)